docs/IR.md

```diff
@@ -1,127 +1,372 @@
+
+### Names Within a Graph
 
 All names MUST adhere to C identifier syntax rules.
 
-#### Nodes
+Names of nodes, inputs, outputs, initializers, and attributes are organized into several namespaces. Within a namespace, each name MUST be unique for each given graph. 
+
+The namespaces are:
 
-Each computation node consists of a name, the identifier of the operator to be invoked, a list of named inputs and outputs, and a list of attribute-value pairs.
+Namespace|Description
+|---|---|
+Attribute|The names of attributes of an operator. Unique for each operator.
+Value|The names of values – node inputs & outputs, tensor values (if named), graph inputs, outputs.
+Node|The names of graph nodes.
+Graph|The names of graphs within a domain, unique within the model domain.
+Operator|The names of operators within a domain.
+Shape|The names of tensor shape variables – scoped to the value information records of a graph, which is where shape variables occur.
+
+
+### Nodes
+
+Computation nodes are comprised of a name, the name of an operator that it invokes, a list of named inputs, a list of named outputs, and a list of attributes. 
+
+Input and outputs are positionally associated with operator inputs and outputs. Attributes are associated with operator attributes by name.
+
+They have the following properties:
+
+Name|Type|Description
+|---|---|---|
+name|string|An optional name of the node, used for diagnostic purposes only.
+input|string[]|Names of the values used by the node to propagate input values to the node operator. It must refer to either a graph input or a node output.
+output|string[]|Names of the outputs used by the node to capture data from the operator invoked by the node. It either introduces a  value in the graph or refers to a graph output.
+op_type|string|The symbolic identifier of the operator to invoke.
+domain|string|The domain of the operator set that contains the operator named by the op_type.
+attribute|Attribute[]|Named attributes, another form of operator parameterization, used for constant values rather than propagated values.
+doc_string|string|A human-readable documentation for this value. Markdown is allowed.
 
 Edges in the computation graph are established by outputs of one node being referenced by name in the inputs of a subsequent node.
 
-Names of tensor values flowing through the graph are unique - a particular name can reference either a graph input or an output of a single node. Reuse of names for several outputs is not allowed.
+The outputs of a given node introduce new names into the graph. The values of node outputs are computed by the node's operator. Node inputs MAY refer to node outputs, graph inputs, and graph initializers. When the name of a node output coincides with the name of a graph output, the graph output's value is the corresponding output value computed by that node.
+
+The graph must use single-static assignment for all node outputs. All node output namess MUST be unique within the graph.
```

Comment:
Can one variable be used as an output for an operator and an input for another operator? If yes, you might want to add  a tiny note.

Reply:
Sorry for causing ambiguation. When looking at BatchNormalization where "mean" appears in its input and output lists, I was considering if using a tensor for both of an operator's input and output is legal. If not legal, I would like to add a sentence to explicitly tell users not to do so.

Reply:
Doing so would, it seems to me, create a cycle in the graph from the node's output to its input, which is explicitly disallowed in the spec.

Reply:
I know it's not allowed because it's a loop, but I would like to emphasize this point.

Reply:
In-place computation violates the single static assignment rule, so that is already covered.

Reply:
"single static assignment" is a magic word to some poor users like me. My background is applied math.